### PR TITLE
feat(cli): add the feature to turn of WRAP for CLI output

### LINKED ISSUE #3336: CLI needs option to truncate field output instead of wrapping it
The new table view in 5.4 is 😍but it causes problems of its own if a field has long values, since the entire value is shown and usually ends up wrapping, which can end up taking many lines of its own - thus making the data almost as unreadable as when it was not rendered as a table. 

This is a mild example, showing how **three** records are rendered over **seven** lines: 

```
ksql> SELECT STARS, CHANNEL, MESSAGE FROM RATINGS ;
+----------------------------+----------------------------+----------------------------+
|STARS                       |CHANNEL                     |MESSAGE                     |
+----------------------------+----------------------------+----------------------------+
|2                           |iOS                         |Surprisingly good, maybe you|
|                            |                            | are getting your mojo back |
|                            |                            |at long last!               |
|1                           |iOS                         |airport refurb looks great, |
|                            |                            |will fly outta here more!   |
|3                           |iOS-test                    |Exceeded all my expectations|
|                            |                            |. Thank you !               |
```

As well as being difficult to read I think it is confusing for the user as it 'feels' wrong to see more lines for fewer data, by default. 

We should seek inspiration from something like sql*plus which has very advanced options on how to configure output (line size, column size, truncate or not, etc). For moment I think there should be a CLI property to enable truncation (which should be set by default, and people can enable wrapping if they want to). The output for the above data would then look like this: 

```
ksql> SELECT STARS, CHANNEL, MESSAGE FROM RATINGS ;
+----------------------------+----------------------------+----------------------------+
|STARS                       |CHANNEL                     |MESSAGE                     |
+----------------------------+----------------------------+----------------------------+
|2                           |iOS                         |Surprisingly good, maybe you[…]|
|1                           |iOS                         |airport refurb looks great, […]|
|3                           |iOS-test                    |Exceeded all my expectations[…]|
```

### PR DESCRIPTION
**Pull request by agavra**
fixes #3336

### Description 

- custom CLI settings are set using Oracle-like `SET CLI PROPERTY VALUE` syntax (which is different from `SET property=value` KSQL syntax)... perhaps this is confusing so I'm open to suggestions
- `SET CLI WRAP ON/OFF` toggles line wrapping (see example below)

```
ksql> SELECT * FROM FOOZ EMIT CHANGES;
+------------------------------------+------------------------------------+------------------------------------+------------------------------------+------------------------------------+
|ROWTIME                             |ROWKEY                              |FOO                                 |BAR                                 |BAZ                                 |
+------------------------------------+------------------------------------+------------------------------------+------------------------------------+------------------------------------+
|1568310477336                       |null                                |hello                               |world                               |no clip                             |
|1568311062781                       |null                                |hello                               |world                               |clip all the things because this is |
|                                    |                                    |                                    |                                    |such a long string!                 |
|1568311085195                       |null                                |what about also clipping something t|world                               |clip all the things because this is |
|                                    |                                    |hat is in the beginning?            |                                    |such a long string!                 |
ksql> SET CLI WRAP OFF;
ksql> SELECT * FROM FOOZ EMIT CHANGES;
+------------------------------------+------------------------------------+------------------------------------+------------------------------------+------------------------------------+
|ROWTIME                             |ROWKEY                              |FOO                                 |BAR                                 |BAZ                                 |
+------------------------------------+------------------------------------+------------------------------------+------------------------------------+------------------------------------+
|1568310477336                       |null                                |hello                               |world                               |no clip                             |
|1568311062781                       |null                                |hello                               |world                               |clip all the things because this ...|
|1568311085195                       |null                                |what about also clipping somethin...|world                               |clip all the things because this ...|
^CQuery terminated
```

The help message:
```
set cli <property> <value>:
	Sets a CLI local property. NOTE that this differs from setting a KSQL property with 'SET property=value' in that it does not affect the server.
```

### TODO

- @JimGalasyn @rmoff - how do you suggest we show all available CLI settings? I think we should be able to link to some documentation that is auto-generated by `CliConfig`, but I'm not sure how to get that working

### Testing done 

- Some unit tests
- E2E CLI testing (see output above)

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by vcrfxia** [comment]
> custom CLI settings are set using Oracle-like SET PROPERTY VALUE syntax (which is different from SET property=value KSQL syntax)... perhaps this is confusing so I'm open to suggestions

My gut says this is confusing. Maybe we can have another command such as `cli-config` to display available CLI properties and their current values, and then use `cli-config set wrap <on/off>` in order to set configs, to make explicit that we're setting a CLI property rather than a KSQL property? I don't love this suggestion either, but I hope we can agree on something since we do have quite a few CLI properties (`output`, `request-pipelining`, and maybe even `server`) so I like the suggestion of centralizing these properties and normalizing how they're set.

**Comment by rmoff** [comment]
Yeah, I think this is going to be confusing:

```
SET 'auto.offset.reset' = 'earliest';
SET WRAP OFF;
```

Do we need to introduce a `SERVER` keyword? 

```
SET SERVER 'auto.offset.reset' = 'earliest';
SET WRAP OFF;
```

Or inverted and probably easier to add on top of what we've got: 

```
SET 'auto.offset.reset' = 'earliest';
SET CLI WRAP OFF;
```

**Comment by agavra** [comment]
thanks for the suggestions @rmoff @vcrfxia! I'm happy with `SET CLI WRAP OFF`, I think it's intuitive and backwards compatible.

**Comment by big-andy-coates** [comment]
Thinking into the future...  we are likely at some point to allow people to set their own date and time formats.  This is pretty common functionality.  

```SET DATEFORMAT yyyy-MM-dd```

What's interesting here is that this would likely be both a client setting, as we'd use it in the CLI to format dates, and also potentially a server setting, i.e. it's passed to the server to allow it to parse supplied date.

So... do we really want to differentiate?

Just throwing that out there for thought! ;)

File: ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java [hunk 1]
```diff
@@ -216,7 +216,9 @@ void handleLine(final String line) throws Exception {
       return;
     }
 
-    handleStatements(line);
+    if (!terminal.maybeHandleCliSpecificCommands(trimmedLine)) {
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java]: Why have you moved this from in `Console`?

I don't see what it's gained you.  You now need to call back into the console, (both here and all tests), rather than just having the console internally handle cli specific commands...

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      );
+
+  public CliConfig(final Map<?, ?> originals) {
+    super(CONFIG_DEF, originals);
+  }
+
+  public CliConfig with(final String property, final Object value) {
+    final Map<String, Object> originals = new HashMap<>(originals());
+    originals.put(property, value);
+    return new CliConfig(originals);
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java]: What sort of error message does the user see if they can the name or value wrong?

A negative test in the Console/Cli for each would be good.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+final class SetCliProperty implements CliSpecificCommand {
+
+  private static final String HELP = "set cli <property> <value>:" + System.lineSeparator()
+      + "\tSets a CLI local property. NOTE that this differs from setting a KSQL "
+      + "property with 'SET property=value' in that it does not affect the server.";
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java]: I think the existing SET requires quoted property name

```suggestion
      + "property with "SET 'property'=value" in that it does not affect the server.";
```

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  public static SetCliProperty create(final BiConsumer<String, String> setProperty) {
+    return new SetCliProperty(setProperty);
+  }
+
+  private SetCliProperty(final BiConsumer<String, String> setProperty) {
+    this.setProperty = setProperty;
+  }
+
+  @Override
+  public boolean matches(final String command) {
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java]: Isn't this the default impl? No need to overwrite.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      + "property with 'SET property=value' in that it does not affect the server.";
+
+  private final BiConsumer<String, String> setProperty;
+
+  public static SetCliProperty create(final BiConsumer<String, String> setProperty) {
+    return new SetCliProperty(setProperty);
+  }
+
+  private SetCliProperty(final BiConsumer<String, String> setProperty) {
+    this.setProperty = setProperty;
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java]: nit: null check

**Comment by agavra** [comment]
Thanks for the review @big-andy-coates!

> LGTM, 'cept I'm left wondering what was gained by moving the maybeHandleCliSpecificCommands call into the CLI aside from more code and having to take the line the console returned and immediately ask that same console if this is actually an internal command it should handle... I don't get it.

Every single time I've looked at this code, I have to dig through and find where we handle the CLI specific commands. I think it makes a lot of sense to have the code be:
1. read line
2. handle line

instead of 
1. read and maybe handle some lines
2. handle the other lines

IMO, whether a command is CLI-specific or not should be part of "handling" not "reading".

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java (hunk 4 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/cmd/SetCliProperty.java]: ah yes - originally it wasn't (to differentiate `SET foo bar` with `SET foo=bar` but I got rid of that distinction

File: ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java (hunk 1 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java]: See my response below - maybe I'm the only one who has been confused by this, but every time I look at this code I get confused.

**Comment by agavra** [comment]
@big-andy-coates - with regards to syntax, I suppose it doesn't matter except that otherwise I need to dig into the contents of the command to figure out whether it's a CLI property or not. That seems... undesirable.

**Comment by big-andy-coates** [comment]
> Thanks for the review @big-andy-coates!
> 
> > LGTM, 'cept I'm left wondering what was gained by moving the maybeHandleCliSpecificCommands call into the CLI aside from more code and having to take the line the console returned and immediately ask that same console if this is actually an internal command it should handle... I don't get it.
> 
> Every single time I've looked at this code, I have to dig through and find where we handle the CLI specific commands. I think it makes a lot of sense to have the code be:
> 
> 1. read line
> 2. handle line
> 
> instead of
> 
> 1. read and maybe handle some lines
> 2. handle the other lines
> 
> IMO, whether a command is CLI-specific or not should be part of "handling" not "reading".

IMHO the change makes the code worse, not better.  As I see it you've changed:

Cli -> get next line to execute from Console
   Console -> read input, handle any internal commands, return any non-internal
Cli -> handle not internal line

To

Cli -> get next line to execute from Console
    Console -> read line and return it
Cli -> call into Console to see it its an internal command and if it is have the console handle it.
      -> otherwise execute it

IMHO this is breaking encapsulation.  Previously only the Console knew about the internal commands.  It dealt with them.  Now the Cli needs to know about them too, but only in so much as to ask the console to handle any of the lines the console just returned. 

To put it another way - previously the Console returned any lines it couldn't handle to the Cli.  Now the Console returns all lines, only for the Cli to have to immediately pass them back to the Console to see if it can handle them.  What's the point?

Maybe if you were to revert this change and then rename `Console.readLine` to `getNextStatement` or something it would make more sense to you?   

I totally agree the architecture of the CLI is questionable.   But I don't think this change helps. It actually couples code more, rather than less.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java]: Did you add a test?  Always nice to reply and let the reviewer know. :D

Comment (big-andy-coates) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java]: Did you add a test? Always nice to reply and let the reviewer know. :D

Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java]: yup - it's in `CliTest`!

**Comment by agavra** [comment]
@big-andy-coates - okay, I will change the readline method back. This was inspired by Oracles SqlPlus: https://docs.oracle.com/cd/B19306_01/server.102/b14357/ch12040.htm where they use `WRAP OFF`:

> Controls whether to truncate the display of a selected row if it is too long for the current line width. OFF truncates the selected row; ON enables the selected row to wrap to the next line.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java (hunk 2 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/CliConfig.java]: EDIT: Moved to `ConsoleTest` after backing out the readline change

**Comment by big-andy-coates** [comment]
> @big-andy-coates - okay, I will change the readline method back. This syntax was inspired by Oracles SqlPlus: https://docs.oracle.com/cd/B19306_01/server.102/b14357/ch12040.htm where they use WRAP OFF:

Fair enough, I wondered if that was the case.

Though... I thought you always looked to Postgres for inspiration - what's with following SqlPlus?!?!? :p

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java [hunk 6]
```diff
@@ -111,14 +120,31 @@ private static void formatRow(
...
     }
   }
 
+  private static boolean shouldClip(final List<String> parts, final int rowsToPrint) {
+    // clip if there are more than one line and any of the remaining lines are non-empty
+    return parts.size() > rowsToPrint
+        && parts.subList(rowsToPrint, parts.size())
+        .stream()
+        .map(String::trim)
+        .noneMatch(String::isEmpty);
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: > // clip if there are more than one line and any of the remaining lines are non-empty

Isn't this check "if there are more than one line and NO LINES ARE EMPTY".  So it would return `false` if there are multiple lines and one is empty.

I think maybe you want '! parts....allMatch(String::isEmpty)'

File: ksql-cli/src/test/java/io/confluent/ksql/cli/CliTest.java [hunk 7]
```diff
@@ -458,6 +458,8 @@ public void testPropertySetUnset() {
     assertRunCommand("unset 'auto.offset.reset'", is(EMPTY_RESULT));
   }
 
+
```
Comment (big-andy-coates) [inline ksql-cli/src/test/java/io/confluent/ksql/cli/CliTest.java]: super nit: blank lines added.

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java (hunk 6 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: good catch! added a test for it
